### PR TITLE
#751 add event bus mechanism and apply it in tc

### PR DESCRIPTION
**Pull request by zhengyangyong**
Signed-off-by: zhengyangyong <yangyong.zheng@qq.com>

https://github.com/seata/seata/issues/751

### CODE REVIEW

File: server/src/test/java/com/alibaba/fescar/server/coordinator/TransactionEventListener.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  public AtomicInteger getCommitEventCount() {
+    return commitEventCount;
+  }
+
+  public AtomicInteger getRollbackEventCount() {
+    return rollbackEventCount;
+  }
+
+  public TransactionEventListener() {
+    this.startEventCount = new AtomicInteger(0);
```
Comment (CoffeeLatte007) [inline server/src/test/java/com/alibaba/fescar/server/coordinator/TransactionEventListener.java]: magic value

File: core/src/main/java/com/alibaba/fescar/core/event/GuavaEventBus.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+public class GuavaEventBus implements EventBus {
+  private final com.google.common.eventbus.EventBus eventBus;
+
+  public GuavaEventBus(String identifier) {
+    this.eventBus = new com.google.common.eventbus.EventBus(identifier);
+  }
+
+  @Override
+  public void register(Object object) {
```
Comment (CoffeeLatte007) [inline core/src/main/java/com/alibaba/fescar/core/event/GuavaEventBus.java]: Why only the test code registers Listener

Comment (zhengyangyong) [reply core/src/main/java/com/alibaba/fescar/core/event/GuavaEventBus.java]: will add listener such as metrics later,not this pr

File: server/src/test/java/com/alibaba/fescar/server/coordinator/TransactionEventListener.java (hunk 1 above)
Comment (zhengyangyong) [reply server/src/test/java/com/alibaba/fescar/server/coordinator/TransactionEventListener.java]: I think may ok in test code ?

**Comment by zhengyangyong** [comment]
> Is th pull request for monitoring ？There are two more points to consider:
> 1.server shutdown or close ,the data will be lost.Whether the data should be persisted and split by day?
> 2.Whether retrycommit, retryRollback and timeout need to monitored

1. because it is a event bus, I think persistence is not necessary
2. yes it is a good suggestion I will try add them

File: server/src/test/java/com/alibaba/fescar/server/coordinator/TransactionEventListener.java (hunk 1 above)
Comment (CoffeeLatte007) [reply server/src/test/java/com/alibaba/fescar/server/coordinator/TransactionEventListener.java]: Ok

File: core/src/main/java/com/alibaba/fescar/core/event/GuavaEventBus.java (hunk 2 above)
Comment (CoffeeLatte007) [reply core/src/main/java/com/alibaba/fescar/core/event/GuavaEventBus.java]: ok

**Comment by zhengyangyong** [comment]
1. Move raise start,commit and rollback event code from DefaultCoordinator to DefaultCore in order to ensure retrycommit and retryRollback monitored
2. Add timeout event in DefaultCoordinator

File: core/src/main/java/io/seata/core/event/GuavaEventBus.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+public class GuavaEventBus implements EventBus {
+  private final com.google.common.eventbus.EventBus eventBus;
+
+  public GuavaEventBus(String identifier) {
+    this.eventBus = new com.google.common.eventbus.EventBus(identifier);
+  }
+
+  @Override
+  public void register(Object object) {
```
Comment (leizhiyuan) [inline core/src/main/java/io/seata/core/event/GuavaEventBus.java]: maybe  Event should be a custom interface， not object， we can limit the type of event。

Comment (zhengyangyong) [reply core/src/main/java/io/seata/core/event/GuavaEventBus.java]: Done

**Comment by leizhiyuan** [comment]
LGTM

**Comment by xingfudeshi** [comment]
@zhengyangyong there are errors need to be solved.

File: server/src/main/java/io/seata/server/event/GlobalTransactionEvent.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package io.seata.server.event;
+
+import io.seata.core.event.Event;
+import io.seata.core.model.GlobalStatus;
+
+public class GlobalTransactionEvent implements Event {
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/event/GlobalTransactionEvent.java]: Is it more appropriate to move to the core module?

File: server/src/test/resources/file.conf [hunk 5]
```diff
@@ -0,0 +1,5 @@
+#reduce delay for test
```
Comment (slievrly) [inline server/src/test/resources/file.conf]: Is it necessary?

File: server/src/test/resources/registry.conf [hunk 6]
```diff
@@ -0,0 +1,7 @@
+config {
```
Comment (slievrly) [inline server/src/test/resources/registry.conf]: Is it necessary?
It seems that it is not consistent with the configuration file of the main project.

File: server/src/main/java/io/seata/server/event/GlobalTransactionEvent.java (hunk 4 above)
Comment (zhengyangyong) [reply server/src/main/java/io/seata/server/event/GlobalTransactionEvent.java]: yes,you are right, Done

File: server/src/test/resources/file.conf (hunk 5 above)
Comment (zhengyangyong) [reply server/src/test/resources/file.conf]: yes ,this file is for test in order to reduce asyn-committing-retry-delay (default is 30 second) and timeout-retry-delay (default is 30 second) , then we can fast check event with `TimeoutRollbacking` status will come from core

File: server/src/test/resources/registry.conf (hunk 6 above)
Comment (zhengyangyong) [reply server/src/test/resources/registry.conf]: same as above

**Comment by zhengyangyong** [comment]
All new file had add author info

**Pull request merged**